subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <net/wifi_mgmt_ext.h>
+#include <net/wifi_credentials.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/net_core.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+
+#include <utils/includes.h>
+#include <utils/common.h>
+#include <common/defs.h>
+#include <wpa_supplicant/config.h>
+#include <wpa_supplicant_i.h>
+#include <driver_i.h>
+
+extern struct k_sem wpa_supplicant_ready_sem;
+extern struct k_mutex wpa_supplicant_mutex;
+extern struct wpa_global *global;
+
+LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
+
+static struct wpa_supplicant *get_wpa_s_handle(const struct device *dev)
+{
+	struct wpa_supplicant *wpa_s = NULL;
+	int ret = k_sem_take(&wpa_supplicant_ready_sem, K_SECONDS(2));
+
+	if (ret) {
+		LOG_ERR("%s: WPA supplicant not ready: %d", __func__, ret);
+		return NULL;
+	}
+
+	k_sem_give(&wpa_supplicant_ready_sem);
+
+	wpa_s = wpa_supplicant_get_iface(global, dev->name);
+
+	if (!wpa_s) {
+		LOG_ERR("%s: Unable to get wpa_s handle for %s", __func__, dev->name);
+		return NULL;
+	}
+
+	return wpa_s;
+}
+
+static void band_to_freq(const struct wifi_credentials_personal *creds, struct wpa_ssid *ssid)
+{
+	/** 2.4 GHz channels: 1-14 */
+	const int channels_2_4ghz[] = {
+		2412, 2417, 2422, 2427,
+		2432, 2437, 2442, 2447,
+		2452, 2457, 2462, 2467,
+		2472, 2484, 0
+	};
+
+	/** 5 GHz channels:
+	 *   36,  40,  44,  48,
+	 *   52,  56,  60,  62,
+	 *   64, 100, 104, 108,
+	 *  112, 116, 120, 124,
+	 *  128, 132, 136, 140,
+	 *  149, 153, 157, 161,
+	 *  165
+	 */
+	const int channels_5ghz[] = {
+		5180, 5200, 5220, 5240,
+		5260, 5280, 5300, 5310,
+		5320, 5500, 5520, 5540,
+		5560, 5580, 5600, 5620,
+		5640, 5660, 5680, 5700,
+		5745, 5765, 5785, 5805,
+		5825, 0
+	};
+
+	if (!(creds->header.flags & (WIFI_CREDENTIALS_FLAG_2_4GHz | WIFI_CREDENTIALS_FLAG_5GHz))) {
+		/* there are no flags indicating band - skip */
+		return;
+	}
+
+	if ((creds->header.flags & (WIFI_CREDENTIALS_FLAG_2_4GHz | WIFI_CREDENTIALS_FLAG_5GHz)) ==
+	    (WIFI_CREDENTIALS_FLAG_2_4GHz | WIFI_CREDENTIALS_FLAG_5GHz)) {
+		/* both bands are indicated - skip */
+		return;
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_2_4GHz) {
+		ssid->freq_list = k_malloc(sizeof(channels_2_4ghz));
```

Comment:
@maxd-nordic Should this be `os_zalloc`?

Reply:
I'll use k_malloc directly here for now. This will be removed entirely when `ctrl_iface` is ready.